### PR TITLE
Add exporting mod lists to text files

### LINKED ISSUE #1152: Mod List Export Options
### Role

I make modpack's and play modded Minecraft frequently 

### Suggestion

Export mod list as .HTML or .TXT

### Benefit

Makes it easier to share your mod list

### This suggestion is unique

- [x] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### Description
I would like to suggest the addition of a quality-of-life feature that allows users to export the mod list for an instance in either a text file or an HTML file format. This feature would greatly enhance the user experience by providing an easy way to save and share the mod list.

#### Text File Example: "PACKNAME-Mods.txt"
```
Example Mod 1 [By Example Dev 1]
Example Mod 2 [By Example Dev 2]
Example Mod 3 [By Example Dev 3]
```

#### HTML File Example: "PACKNAME-Mods.html"
```html
<ul>
   <li><a href="https://modrinth.com/mod/example-mod-1">Example Mod 1 [By Example Dev 1]</a></li>
   <li><a href="https://modrinth.com/mod/example-mod-2">Example Mod 2 [By Example Dev 2]</a></li>
   <li><a href="https://www.curseforge.com/minecraft/mc-mods/example-mod-3">Example Mod 3 [By Example Dev 3]</a></li>
</ul>
```

### Benefits

- Easy Sharing: Users can share the exported mod lists with others, making it simpler to discuss and compare mods.
- Documentation: Exported mod lists can serve as documentation for reference or troubleshooting purposes.
---

## Image Concepts

Please find the image concepts below:

[](https://postimg.cc/nXZDnWJF)
[](https://postimg.cc/87Hrcbph)
[](https://postimg.cc/bs1nch4V)

These image concepts provide visual representations of the ideas and suggestions discussed. They can help the development team better understand and implement the requested features.

### Thank you for reading, I hope to see this implemented in the future.

### PR DESCRIPTION
**Pull request by Trial97**
fixes #1152
This has just the mods to text (~~no curseforge export for the moment~~ that's implemented in another [PR](https://github.com/PrismLauncher/PrismLauncher/pull/1249) )

### CODE REVIEW

**Comment by DioEgizio** [comment]
few issues:
- image this :P
- i think that it should be renamed from "text" to "Mod List" because all you're exporting is a mod list
- should probably support plaintext too
- it should have a save button where to save the html/md/plaintext
- should probably also render the md and the html somewhere

File: launcher/ui/MainWindow.ui [hunk 1]
```diff
@@ -479,6 +479,14 @@
     <string>Modrinth (mrpack)</string>
    </property>
   </action>
+  <action name="actionExportInstanceToModList">
+   <property name="icon">
+    <iconset theme="new"/>
+   </property>
+   <property name="text">
+    <string>ModList (txt)</string>
```
Comment (DioEgizio) [inline launcher/ui/MainWindow.ui]:
```suggestion
    <string>Mod List</string>
```

File: launcher/ui/dialogs/ExportToModListDialog.cpp [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    if (ui->versionCheckBox->isChecked())
+        opt |= ExportToModList::Version;
+    if (ui->urlCheckBox->isChecked())
+        opt |= ExportToModList::Url;
+    auto txt = ExportToModList::ExportToModList(m_allMods, format, static_cast<ExportToModList::OptionalData>(opt));
+    ui->finalText->setPlainText(txt);
+    QString exampleLine;
+    switch (format) {
+        case ExportToModList::HTML: {
+            exampleLine = "<ul><a href=\"{url}\">{name}</a>[{version}] by {authors}</ul>";
```
Comment (DioEgizio) [inline launcher/ui/dialogs/ExportToModListDialog.cpp]:
```suggestion
            exampleLine = "<ul><a href=\"{url}\">{name}</a> [{version}] by {authors}</ul>";
```

File: launcher/ui/dialogs/ExportToModListDialog.cpp [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+            ui->resultText->setHtml(txt);
+            break;
+        }
+        case ExportToModList::MARKDOWN: {
+            exampleLine = "[{name}]({url})[{version}] by {authors}";
```
Comment (DioEgizio) [inline launcher/ui/dialogs/ExportToModListDialog.cpp]:
```suggestion
            exampleLine = "[{name}]({url}) [{version}] by {authors}";
```

File: launcher/ui/dialogs/ExportToModListDialog.cpp [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+            ui->resultText->setHtml(markdownToHTML(txt));
+            break;
+        }
+        case ExportToModList::PLAINTXT: {
+            exampleLine = "name: {name}; url: {url}; version: {version}; authors: {authors};";
```
Comment (DioEgizio) [inline launcher/ui/dialogs/ExportToModListDialog.cpp]:
```suggestion
            exampleLine = "{name} ({url}) [{version}] by {authors}";
```

**Review by DioEgizio — COMMENTED** [review]
image
also the html renderer doesn't seem to correctly linebreak

**Comment by Scrumplex** [comment]
This is a weird feature. I don't really see the use case if people can just create mrpacks now anyway.

**Comment by DioEgizio** [comment]
I can see the use of it tbh. You don't want an entire modpack for just listing your mods. It can be useful also for example to modpack creators that want to list their mod in their description. With this they can just export- mod list- markdown and they can just copy paste it (same for html on CF). I can see many uses. Not to mention that some mods are on cf and some on modrinth

**Comment by Trial97** [comment]
If not desired feel free to close this but also close the linked issue with an explanation of why this is not wanted

**Comment by DioEgizio** [comment]
I think this is useful in many scenarios tbh

**Comment by Scrumplex** [comment]
Well. This is something that can be automated using shell scripts. Like, I don't see how this feature fits in the context of a launcher.

**Comment by DioEgizio** [comment]
Shell scripts parsing metadata information?

**Comment by Scrumplex** [comment]
> Shell scripts parsing metadata information?

You can use tools like `yq` to read the Packwiz TOML files.

> It can be useful also for example to modpack creators that want to list their mod in their description

I haven't seen many modpacks do that (usually because the mod platform will already do it for you mostly)

---

I just don't think this feature belongs into the launcher. I really hate this word, but I think "bloat" fits well here. If we add niche features like this, we only increase the amount of maintenance needed in the future.

Also, where does this feature end? Are we gonna need an HTML and a BBCode export as well? Maybe someone needs to post the modlist in a Minecraft Forums post, which doesn't support Markdown

**Comment by Trial97** [comment]
> Also, where does this feature end? Are we gonna need an HTML and a BBCode export as well? Maybe someone needs to post the modlist in a Minecraft Forums post, which doesn't support Markdown

Well, this already supports HTML and for other stuff, I added a custom format that lets the user define how each line should look.

**Comment by DioEgizio** [comment]
> You can use tools like yq to read the Packwiz TOML files.

It wouldn't need just to to that, but to read the metadata inside the mod file. At that point it's not even a shell scripts, it's a mod manager

> I haven't seen many modpacks do that (usually because the mod platform will already do it for you mostly)

Some do

**Comment by Scrumplex** [comment]
Look. I usually try to look at the problem in an issue, instead of accepting their proposed solution to fix their problem. The issue states the following benefits:

> - Easy Sharing: Users can share the exported mod lists with others, making it simpler to discuss and compare mods.
> - Documentation: Exported mod lists can serve as documentation for reference or troubleshooting purposes.

Both of these can be achieved using the mrpack export. So why add another feature for the same purpose?

**Comment by DioEgizio** [comment]
mrpack export can only be used for, well, modrinth mods and mods that can be redistributed. It doesn't work for CF only mods and feels redundant for just a mod list. Like imagine if someone asks you what mods you use with the links to them and you literally have to make an entire modpack. Not to mention you also might want to add the mods to an already existing instance

**Comment by Scrumplex** [comment]
My problem with this is that this is intended to output to a human-readable format. You can't use an exported mod list to import it into another instance anyway. It would be a lot of manual work.

If you want to share your mods with a friend you could send them the list of the jar files, you could take a screenshot of your mods either in Prism or even in-game.

**Comment by DioEgizio** [comment]
Still, I don't understand what's the problem with a human readable mod list. It's the reason why things like moddermore exist after all

**Comment by Scrumplex** [comment]
The reasons against are 700 additional lines of code that have to be maintained.

**Comment by DioEgizio** [comment]
I mean it's not like this is the only feature that could be considered niche (but I don't think it is tbh). We literally have a java profiler option

**Comment by Trial97** [comment]
Now this would also implement the export options mentioned in the following issues https://github.com/PrismLauncher/PrismLauncher/issues/214, https://github.com/PrismLauncher/PrismLauncher/issues/1178 (no importing for this)
I'm linking them here so that they can be addressed once this PR is closed.

**Review by DioEgizio — APPROVED** [review]

**Comment by TheKodeToad** [comment]
This is quite nice!

Maybe it should prompt to update/"sync" the mods?
I found a lot of URLs were missing...

**Comment by TheKodeToad** [comment]
Ohh - also maybe `Optional Info` could change into buttons to insert the placeholders on the Custom option.

And instead of preventing the Template from being edited, you could detect when it's edited and switch to custom...

**Comment by Trial97** [comment]
Thanks for the feedback:
- instead of prompting to update mods, I added a message that should explain that the metadata is needed
- updated the template behavior 
- regarding Oprional info I can't really picture what you mean so I kept it as is

**Comment by TheKodeToad** [comment]
Maybe have buttons to insert {author} etc...

**Comment by Trial97** [comment]
I don't know what to say about this...
If I go this route, I should remove the extra file types as they do not make sense anymore if the buttons change the template.
In that case, I would only have a custom and a template, is that what you want?

**Comment by TheKodeToad** [comment]
Oops... no, I meant only have that behaviour in custom

**Comment by Trial97** [comment]
This should solve it then

**Comment by TheKodeToad** [comment]
Nice :3

I'll try to properly review the changes now if i don't get distracted yet again :P

File: launcher/modplatform/helpers/ExportToModList.cpp [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+            else if (mod->authors().length() > 1)
+                authors = QString("\"%1\"").arg(mod->authors().join(","));
+            data << authors;
+        }
+        lines << data.join(",");
+    }
+    return lines.join("\n");
+}
+
+QString ExportToModList(QList<Mod*> mods, Formats format, OptionalData extraData)
```
Comment (TheKodeToad) [inline launcher/modplatform/helpers/ExportToModList.cpp]:
```suggestion
QString exportToModList(QList<Mod*> mods, Formats format, OptionalData extraData)
```

File: launcher/modplatform/helpers/ExportToModList.cpp [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+            return toJSON(mods, extraData);
+        case CSV:
+            return toCSV(mods, extraData);
+        default: {
+            return QString("unknown format:%1").arg(format);
+        }
+    }
+}
+
+QString ExportToModList(QList<Mod*> mods, QString lineTemplate)
```
Comment (TheKodeToad) [inline launcher/modplatform/helpers/ExportToModList.cpp]:
```suggestion
QString exportToModList(QList<Mod*> mods, QString lineTemplate)
```

File: launcher/modplatform/helpers/ExportToModList.cpp [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+                line += QString(" [%1]").arg(ver.toHtmlEscaped());
+        }
+        if (extraData & Authors && !mod->authors().isEmpty())
+            line += " by " + mod->authors().join(", ").toHtmlEscaped();
+        lines.append(QString("<li>%1</li>").arg(line));
+    }
+    return QString("<html><body><ul>\n\t%1\n</ul></body></html>").arg(lines.join("\n\t"));
+}
+
+QString toMARKDOWN(QList<Mod*> mods, OptionalData extraData)
```
Comment (TheKodeToad) [inline launcher/modplatform/helpers/ExportToModList.cpp]: ARGHHHHHHHHHH!!!!!!!!!!

I think Markdown would make more sense :laughing:

File: launcher/modplatform/helpers/ExportToModList.cpp [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+                line += QString(" [%1]").arg(ver);
+        }
+        if (extraData & Authors && !mod->authors().isEmpty())
+            line += " by " + mod->authors().join(", ");
+        lines << "- " + line;
+    }
+    return lines.join("\n");
+}
+
+QString toPLAINTXT(QList<Mod*> mods, OptionalData extraData)
```
Comment (TheKodeToad) [inline launcher/modplatform/helpers/ExportToModList.cpp]: PlainTXT?

**Review by TheKodeToad — CHANGES_REQUESTED** [review]
I'm confused what the changes to dependencies are for...

**Comment by Trial97** [comment]
> I'm confused what the changes to dependencies are for...

oops I think I merged the wrong develop(I will remove that)

**Comment by Trial97** [comment]
Ok now it should be merged ok

File: launcher/ui/dialogs/ExportToModListDialog.ui [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+          </property>
+         </widget>
+        </item>
+       </layout>
+      </widget>
+     </item>
+     <item>
+      <widget class="QLabel" name="warningLabel">
+       <property name="text">
+        <string>This depends on the mods meta data. To ensure the meta data run at least one time the mods update on the selected instance(no need to update the mods).</string>
```
Comment (TheKodeToad) [inline launcher/ui/dialogs/ExportToModListDialog.ui]:
```suggestion
        <string>This depends on the mods' metadata. To ensure it is available, run an update on the instance. Installing the updates isn't necessary.</string>
```

**Review by TheKodeToad — APPROVED** [review]
good! ty

**Review by Scrumplex — APPROVED** [review]

**Comment by Scrumplex** [comment]
I still think this is an overkill feature, but eh

**Comment by ruguysgoingtrickortreating** [comment]
Hi, I stumbled across this pull request because my enhancement was mentioned, so would it be possible to import the file this generates and automatically download the mods from their respective repositories like the download mods menu? It would add a lot of use to this feature.

**Comment by Scrumplex** [comment]
> Hi, I stumbled across this pull request because my enhancement was mentioned, so would it be possible to import the file this generates and automatically download the mods from their respective repositories like the download mods menu? It would add a lot of use to this feature.

You can already do this using the Modrinth export feature. Just export your instance as a Modrinth modpack and import it somewhere else.

**Comment by DioEgizio** [comment]
Same with the cf export but I think he's talking about when mods have different types providers

**Pull request merged**